src/init.cpp

```diff
@@ -1707,15 +1707,33 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     // Start the thread that notifies listeners of transactions that have been
     // recently added to the mempool, or have been added to or removed from the
     // chain. We perform this before step 10 (import blocks) so that the
-    // original value of chainActive.Tip(), which corresponds with the wallet's
-    // view of the chaintip, is passed to ThreadNotifyWallets before the chain
-    // tip changes again.
+    // original value of chainActive.Tip() can be passed to ThreadNotifyWallets
+    // before the chain tip changes again.
     {
         CBlockIndex *pindexLastTip;
         {
             LOCK(cs_main);
             pindexLastTip = chainActive.Tip();
         }
+
+        // However, if a wallet is enabled, we actually want to start notifying
+        // from the block which corresponds with the wallet's view of the chain
+        // tip. In particular, we want to handle the case where the node shuts
+        // down uncleanly, and on restart the chain's tip is potentially up to
+        // an hour of chain sync older than the wallet's tip. We assume here
+        // that there is only a single wallet connected to the validation
+        // interface, which is currently true.
+#ifdef ENABLE_WALLET
+        if (pwalletMain)
+        {
+            LOCK2(cs_main, pwalletMain->cs_wallet);
+            const auto walletBestBlock = pwalletMain->GetBestBlock();
+            if (walletBestBlock != nullptr) {
+                pindexLastTip = walletBestBlock;
```

Comment:
In looking at uses of `pindexLastTip`, I think that we are missing some Orchard merkle tree handling here: https://github.com/zcash/zcash/pull/5809/files#diff-cd341be4887e632ceba211aee03c948d964396f7aed0146435163844fbcbad42R164-L154

Reply:
@nuttycom I can't tell what you're linking to, that link is bad or something.

Reply:
Sorry, here's a better link: https://github.com/zcash/zcash/blob/nu5-consensus/src/validationinterface.cpp#L139-L154

Reply:
Aha, I knew I'd seen it: we worked on it in a pairing for your PR #5789.

Reply:
Good grief, I clearly am losing it.